ParseError with try / catch expressions

I ran into a syntax edge case that triggers a ParseError:

Scala: 3.3.4
sbt-scalafix: 0.14.0
sbt-scalafmt: 2.5.4

**.scalafix.conf**
```
DisableSyntax.noVars = true
DisableSyntax.noThrows = true
DisableSyntax.noNulls = true
DisableSyntax.noReturns = true
DisableSyntax.noWhileLoops = true
DisableSyntax.noAsInstanceOf = true
DisableSyntax.noIsInstanceOf = true
DisableSyntax.noXml = true
OrganizeImports.expandRelative = true
OrganizeImports.removeUnused = true
OrganizeImports.targetDialect = Auto
rules = [DisableSyntax, LeakingImplicitClassVal, NoAutoTupling, NoValInForComprehension, OrganizeImports, RedundantSyntax, RemoveUnused]
```

**.scalafmt.conf**
```
maxColumn = 120
runner.dialect = scala3
assumeStandardLibraryStripMargin = true
rewrite.imports.sort = original
project.excludePaths = ["glob:**/metals.sbt"]
version = 3.8.6
rewrite.rules = [Imports, SortModifiers]
```

Snippet:
```scala
def test(a: Int, b: String): Unit = {}

val foo = test(
  try 1
  catch case _ => 2,
  "bar"
)
```
scalac is happy about that, but scalafix complains:
```
[error] ./package.scala:25:20: error: `;` expected but `,` found
[error]   catch case _ => 2,
[error]                    ^               ^
[error] (skunk / Compile / scalafix) scalafix.sbt.ScalafixFailed: ParseError
```
scalafmt fails as well:
```
[error] (skunk / Compile / scalafmt) org.scalafmt.sbt.ScalafmtSbtReporter$ScalafmtSbtError: scalafmt: ./package.scala:25: error: [dialect scala3] `;` expected but `,` found
[error]   catch case _ => 2,
[error]                    ^ [./package.scala]
```

I initially opened this issue in scalafix: https://github.com/scalacenter/scalafix/issues/2181

--- a/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScannerTokens.scala
+++ b/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScannerTokens.scala
@@ -535,17 +535,17 @@ final class ScannerTokens(val tokens: Tokens)(implicit dialect: Dialect) {
           case _ :: rs => OutdentInfo(null, rs)
           case _ => null
         }
-      case _: LeftBracket => currRef(RegionBracket :: sepRegions)
+      case _: LeftBracket => currRef(RegionBracket(nextIndent) :: sepRegions)
       case _: RightBracket => mkOutdents(sepRegions) {
           case (r: SepRegionIndented) :: rs => OutdentInfo(r, rs)
-          case RegionBracket :: rs => OutdentInfo(null, rs, done = true)
+          case (_: RegionBracket) :: rs => OutdentInfo(null, rs, done = true)
           case _ :: rs => OutdentInfo(null, rs)
           case _ => null
         }
-      case _: LeftParen => currRef(RegionParen :: sepRegions)
+      case _: LeftParen => currRef(RegionParen(nextIndent) :: sepRegions)
       case _: RightParen => mkOutdents(sepRegions) {
           case (r: SepRegionIndented) :: rs => OutdentInfo(r, rs)
-          case RegionParen :: rs => OutdentInfo(null, rs, true)
+          case (_: RegionParen) :: rs => OutdentInfo(null, rs, true)
           case _ :: rs => OutdentInfo(null, rs)
           case _ => null
         }
@@ -633,8 +633,8 @@ final class ScannerTokens(val tokens: Tokens)(implicit dialect: Dialect) {
           case r: RegionFor if r.isClosingConditionToken(prev) => rs
           case _ => if (prevToken.is[AtEOL] || curr.is[CloseDelim]) rs else regions
         }
-      case RegionParen :: RegionForMaybeParens :: rs if curr.is[LeftArrow] =>
-        RegionParen :: RegionForParens :: rs
+      case (r: RegionParen) :: RegionForMaybeParens :: rs if curr.is[LeftArrow] =>
+        r :: RegionForParens :: rs
       case _ => regions
     }
 
@@ -1018,7 +1018,7 @@ object ScannerTokens {
   @tailrec
   private def inParens(regions: List[SepRegion]): Boolean = regions.nonEmpty &&
     (regions.head match {
-      case r: RegionDelimNonIndented => r eq RegionParen
+      case r: RegionDelimNonIndented => r.isInstanceOf[RegionParen]
       case _ => inParens(regions.tail)
     })
 

--- a/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SepRegion.scala
+++ b/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/SepRegion.scala
@@ -28,8 +28,8 @@ case class RegionIndent(override val indent: Int) extends SepRegionIndented with
 
 case class RegionLine(override val indent: Int) extends RegionNonDelimNonIndented with CanProduceLF
 
-case object RegionParen extends RegionDelimNonIndented
-case object RegionBracket extends RegionDelimNonIndented
+case class RegionParen(override val indent: Int) extends RegionDelimNonIndented
+case class RegionBracket(override val indent: Int) extends RegionDelimNonIndented
 case class RegionBrace(override val indent: Int) extends RegionDelimNonIndented with CanProduceLF
 
 case object RegionCaseMark extends RegionNonDelimNonIndented

--- a/tests/shared/src/test/scala/scala/meta/tests/parsers/DefnSuite.scala
+++ b/tests/shared/src/test/scala/scala/meta/tests/parsers/DefnSuite.scala
@@ -407,4 +407,30 @@ class DefnSuite extends ParseSuite {
     runTestAssert[Stat](code, layout)(tree)
   }
 
+  test("#4177") {
+    val code =
+      """|val foo = test(
+         |  try 1
+         |  catch { case _ => 2 },
+         |  "bar"
+         |)
+         |""".stripMargin
+    val layout =
+      """|val foo = test(try 1 catch {
+         |  case _ => 2
+         |}, "bar")
+         |""".stripMargin
+    val tree = Defn.Val(
+      Nil,
+      List(patvar("foo")),
+      None,
+      tapply(
+        "test",
+        Term.Try(lit(1), Some(Term.CasesBlock(List(Case(patwildcard, None, lit(2))))), None),
+        lit("bar")
+      )
+    )
+    runTestAssert[Stat](code, layout)(tree)
+  }
+
 }

--- a/tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/ControlSyntaxSuite.scala
+++ b/tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/ControlSyntaxSuite.scala
@@ -4166,11 +4166,22 @@ class ControlSyntaxSuite extends BaseDottySuite {
          |  "bar"
          |)
          |""".stripMargin
-    val error =
-      """|<input>:3: error: `;` expected but `,` found
-         |  catch case _ => 2,
-         |                   ^""".stripMargin
-    runTestError[Stat](code, error)
+    val layout =
+      """|val foo = test(try 1 catch {
+         |  case _ => 2
+         |}, "bar")
+         |""".stripMargin
+    val tree = Defn.Val(
+      Nil,
+      List(patvar("foo")),
+      None,
+      tapply(
+        "test",
+        Term.Try(lit(1), Some(Term.CasesBlock(List(Case(patwildcard, None, lit(2))))), None),
+        lit("bar")
+      )
+    )
+    runTestAssert[Stat](code, layout)(tree)
   }
 
 }

--- a/tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/ControlSyntaxSuite.scala
+++ b/tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/ControlSyntaxSuite.scala
@@ -4158,4 +4158,19 @@ class ControlSyntaxSuite extends BaseDottySuite {
     runTestAssert[Stat](code, layout)(tree)
   }
 
+  test("#4177") {
+    val code =
+      """|val foo = test(
+         |  try 1
+         |  catch case _ => 2,
+         |  "bar"
+         |)
+         |""".stripMargin
+    val error =
+      """|<input>:3: error: `;` expected but `,` found
+         |  catch case _ => 2,
+         |                   ^""".stripMargin
+    runTestError[Stat](code, error)
+  }
+
 }

